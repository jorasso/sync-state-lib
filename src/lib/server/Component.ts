import { GetObjectProperty } from "../utils/GetObjectProperty"

export enum ComponentChangeType {
  "complete" = 1,
  "partial" = 2,
}

interface ComponentChange {
  isComponent: true
  from: Component
  to: Component
  meta?: any
}

interface PrimitiveChange {
  isComponent: false
  from: number | string | boolean
  to: number | string | boolean
  meta?: any
}

export class Component {
  readonly isComponent: boolean = true

  static AllSyncedAlphabeticalOrder: string[] = []
  static AllSynced: string[] = []
  static SyncedComponents: string[] = []
  static SyncedProperties: string[] = []

  // This way we build a tree of components
  private parentComponent!: Component

  // Both things are set by the parent component.
  private parentNumId!: number
  private parentPropertyKey!: string

  // Child components
  protected components = new Map<Component, number>()

  // Observed properties
  protected properties = new Map<string, number>()

  protected primitivePropertiesChanges: Map<number, PrimitiveChange> = new Map()
  protected childComponentsChanges: Map<number, ComponentChange> = new Map()

  protected changes: (PrimitiveChange | ComponentChange)[] = []

  protected lastNumId: number = 0

  private updateReportedToParent: boolean = false

  private numIdToOrder: number[] = []
  private orderToNumId: number[] = []

  private allPropertyKeys: string[] = []

  getNumId() {
    const numId = this.lastNumId ?? 0

    this.lastNumId += 1

    return numId
  }

  makeSynced() {
    const properties: string[] =
      GetObjectProperty(this.constructor, "SyncedProperties") ?? []
    const components: string[] =
      GetObjectProperty(this.constructor, "SyncedComponents") ?? []
    const all: string[] = GetObjectProperty(this.constructor, "AllSynced") ?? []
    const allOrder: string[] = [...all].sort()

    this.allPropertyKeys = allOrder

    all.forEach((propertyKey) => {
      const isProperty = properties.includes(propertyKey)
      const numId = allOrder.indexOf(propertyKey)

      const order = this.getNumId()

      if (isProperty) {
        let value = GetObjectProperty(this, propertyKey)

        this.numIdToOrder[numId] = order
        this.orderToNumId[order] = numId

        this.properties.set(propertyKey, numId)

        Object.defineProperty(this, propertyKey, {
          get: () => {
            return value
          },
          set: (newValue: any) => {
            this.addPrimitivePropertyChange(numId, value, newValue)

            value = newValue
          },
          enumerable: true,
          configurable: true,
        })
      } else {
        let component = GetObjectProperty(this, propertyKey)

        this.numIdToOrder[numId] = order
        this.orderToNumId[order] = numId

        this.components.set(component, numId)

        if (component?.isComponent) {
          component.setParent(this, propertyKey, numId)
        }

        Object.defineProperty(this, propertyKey, {
          get: () => {
            return component
          },
          set: (newComponent: Component) => {
            const numId = component?.parentNumId ?? this.getNumId()

            if (component) {
              component.dispose()
            }

            if (newComponent) {
              newComponent.setParent(this, propertyKey, numId)
            }

            this.addChildComponentPropertyChange(numId, component, newComponent)

            component = newComponent
          },
          enumerable: true,
          configurable: true,
        })
      }
    })
  }

  makeSyncedBKP() {
    const properties: string[] =
      GetObjectProperty(this.constructor, "SyncedProperties") ?? []
    const components: string[] =
      GetObjectProperty(this.constructor, "SyncedComponents") ?? []
    const all: string[] = GetObjectProperty(this.constructor, "AllSynced") ?? []

    properties.forEach((propertyKey) => {
      let value = GetObjectProperty(this, propertyKey)
      let numId = this.getNumId()

      this.properties.set(propertyKey, numId)

      Object.defineProperty(this, propertyKey, {
        get: () => {
          return value
        },
        set: (newValue: any) => {
          this.addPrimitivePropertyChange(numId, value, newValue)

          value = newValue
        },
        enumerable: true,
        configurable: true,
      })
    })

    components.forEach((propertyKey) => {
      let component = GetObjectProperty(this, propertyKey)

      const numId = this.getNumId()
      this.components.set(component, numId)

      if (component?.isComponent) {
        component.setParent(this, propertyKey, numId)
      }

      Object.defineProperty(this, propertyKey, {
        get: () => {
          return component
        },
        set: (newComponent: Component) => {
          const numId = component?.parentNumId ?? this.getNumId()

          if (component) {
            component.dispose()
          }

          if (newComponent) {
            newComponent.setParent(this, propertyKey, numId)
          }

          this.addChildComponentPropertyChange(numId, component, newComponent)

          component = newComponent
        },
        enumerable: true,
        configurable: true,
      })
    })
  }

  // If the value of a property changes, or the property that holds a component is reassigned
  addPrimitivePropertyChange = (
    numId: number,
    oldValue: any,
    newValue: any,
    meta?: any
  ) => {
    if (oldValue === newValue) {
      return
    }

    let change = this.primitivePropertiesChanges.get(numId)

    if (change) {
      change.to = newValue
    } else {
      change = {
        isComponent: false,
        from: oldValue,
        to: newValue,
      }

      this.primitivePropertiesChanges.set(numId, change)
    }

    this.changes[this.numIdToOrder[numId]] = change

    if (this.parentComponent && !this.updateReportedToParent) {
      this.updateReportedToParent = true
      this.parentComponent.addChildComponentUpdatedChange(this)
    }

    if (meta) {
      change.meta = meta
    }
  }

  addChildComponentPropertyChange = (
    numId: number,
    currentComponent: Component,
    newComponent: Component,
    meta?: any
  ) => {
    if (currentComponent) {
      this.components.delete(currentComponent)
    }

    if (newComponent) {
      this.components.set(newComponent, numId)
    }

    let change = this.childComponentsChanges.get(numId)

    if (change) {
      change.to = newComponent
    } else {
      change = {
        isComponent: true,
        from: currentComponent,
        to: newComponent,
      }
      this.childComponentsChanges.set(numId, change)
    }

    if (meta) {
      change.meta = meta
    }

    this.changes[this.numIdToOrder[numId]] = change

    if (this.parentComponent && !this.updateReportedToParent) {
      this.updateReportedToParent = true
      this.parentComponent.addChildComponentUpdatedChange(this)
    }
  }

  addChildComponentUpdatedChange = (childComponent: Component) => {
    const numId = childComponent.getParentNumId()
    let change = this.childComponentsChanges.get(numId)

    if (change) {
      change.to = childComponent
    } else {
      change = {
        isComponent: true,
        from: childComponent,
        to: childComponent,
      }
      this.childComponentsChanges.set(numId, change)
    }

    this.changes[this.numIdToOrder[numId]] = change

    if (this.parentComponent && !this.updateReportedToParent) {
      this.updateReportedToParent = true
      this.parentComponent.addChildComponentUpdatedChange(this)
    }
  }

  setParent(
    parentComponent: Component,
    parentPropertyKey: string,
    parentNumId: number
  ) {
    this.parentComponent = parentComponent
    this.parentPropertyKey = parentPropertyKey
    this.parentNumId = parentNumId
  }

  getParent() {
    return this.parentComponent
  }

  getParentPropertyKey() {
    return this.parentPropertyKey
  }

  getParentNumId() {
    return this.parentNumId
  }

  dispose() {
    this.parentComponent = undefined
    this.parentPropertyKey = undefined
    this.parentNumId = undefined
  }

  getCompleteState() {
    const state: any = [ComponentChangeType.complete]

    this.orderToNumId.forEach((numId, order) => {
      const propertyKey = this.allPropertyKeys[numId]

      const isProperty = this.properties.has(propertyKey)

      if (isProperty) {
        const property = GetObjectProperty(this, propertyKey)

        state.push(numId)
        state.push(property)
      } else {
        const component = GetObjectProperty(this, propertyKey) as Component

        if (component) {
          state.push(numId)
          state.push(component.getCompleteState())
        }
      }
    })

    return state
  }

  getCompleteStateBKP() {
    const state: any = [ComponentChangeType.complete]

    state.push(this.properties.size)

    this.properties.forEach((numId, key) => {
      state.push(key)
      state.push(numId)

      state.push(GetObjectProperty(this, key))
    })

    this.components.forEach((numId, component) => {
      state.push(component.getParentPropertyKey())
      state.push(numId)
      state.push(component.getCompleteState())
    })

    return state
  }

  getRecentChanges() {
    const changes: any = [ComponentChangeType.partial]

    this.changes.forEach((change, order) => {
      const numId = this.orderToNumId[order]

      if (change.isComponent) {
        if (change.from === change.to) {
          const recentChanges = change.to.getRecentChanges()

          // Don't include empty changes (only contains the type of change as partial, and no actual changes)
          if (recentChanges.length > 1) {
            changes.push(numId)
            changes.push(recentChanges)
          }
        } else {
          changes.push(numId)
          changes.push(change.to.getCompleteState())
        }
      } else {
        if (change.from === change.to) {
          return
        }
        changes.push(numId)
        changes.push(change.to)
      }
    })

    return changes
  }

  getRecentChangesBKP() {
    const changes: any = [ComponentChangeType.partial]

    this.primitivePropertiesChanges.forEach((change, numId) => {
      if (change.from === change.to) {
        return
      }
      changes.push(numId)
      changes.push(change.to)
    })

    this.childComponentsChanges.forEach((change, numId) => {
      if (change.from === change.to) {
        const recentChanges = change.to.getRecentChanges()

        // Don't include empty changes (only contains the type of change as partial, and no actual changes)
        if (recentChanges.length > 1) {
          changes.push(numId)
          changes.push(change.to.getRecentChanges())
        }
      } else {
        changes.push(numId)
        changes.push(change.to.getCompleteState())
      }
    })

    return changes
  }

  resetRecentChanges() {
    this.changes = []

    this.updateReportedToParent = false

    this.primitivePropertiesChanges.clear()
    this.childComponentsChanges.clear()

    this.components.forEach((_, component) => {
      component.resetRecentChanges()
    })
  }
}
