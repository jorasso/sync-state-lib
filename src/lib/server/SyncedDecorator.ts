// Decorates the value of a property to be synced with the client
export const Synced = () => {
  return function (target: any, propertyKey: string) {
    // Initialize the static property if it doesn't exist
    if (!target.constructor.hasOwnProperty("SyncedProperties")) {
      target.constructor.SyncedProperties = []
    }

    if (!target.constructor.hasOwnProperty("AllSynced")) {
      target.constructor.AllSynced = []
    }

    // Add the property key to the list of synced properties
    target.constructor.SyncedProperties.push(propertyKey)
    target.constructor.AllSynced.push(propertyKey)
  }
}

// Decorates the value of a property to be a component synced with the client
export const SyncedComponent = () => {
  return function (target: any, propertyKey: string) {
    // Initialize the static property if it doesn't exist
    if (!target.constructor.hasOwnProperty("SyncedComponents")) {
      target.constructor.SyncedComponents = []
    }

    if (!target.constructor.hasOwnProperty("AllSynced")) {
      target.constructor.AllSynced = []
    }

    // Add the property key to the list of synced components
    target.constructor.SyncedComponents.push(propertyKey)
    target.constructor.AllSynced.push(propertyKey)
  }
}

// Decorates a class to be synced with the client.
// This is needed to avoid having to call makeSynced() manually.
// Otherwise defineProperty() doesn't work as expected with derived classes.
export const WithSynced = <T extends { new (...args: any[]): {} }>(
  constructor: T
) => {
  return class extends constructor {
    constructor(...args: any[]) {
      super(...args)
      //@ts-ignore
      if (typeof this["makeSynced"] === "function") {
        //@ts-ignore
        this["makeSynced"]()
      }
    }
  }
}
