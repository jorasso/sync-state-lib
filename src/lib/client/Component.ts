import { ComponentChangeType } from "../server/Component"
import { GetObjectProperty } from "../utils/GetObjectProperty"
import { SetObjectProperty } from "../utils/SetObjectProperty"

export class Component {
  static AllSyncedAlphabeticalOrder: string[] = []
  static AllSynced: string[] = []
  static SyncedComponents: string[] = []
  static SyncedProperties: string[] = []

  protected components = new Map<number, Component>()

  protected properties = new Map<number, string>()

  private listeners = new Map<string, Array<(value: any) => void>>()

  private parentComponent!: Component
  private parentNumId!: number
  private parentPropertyKey!: string

  makeSynced() {
    const properties: string[] =
      GetObjectProperty(this.constructor, "SyncedProperties") ?? []
    const components: string[] =
      GetObjectProperty(this.constructor, "SyncedComponents") ?? []
    const all: string[] = GetObjectProperty(this.constructor, "AllSynced") ?? []
    const allOrder: string[] = [...all].sort()

    all.forEach((propertyKey) => {
      const isProperty = properties.includes(propertyKey)
      const numId = allOrder.indexOf(propertyKey)

      if (isProperty) {
        this.properties.set(numId, propertyKey)
      } else {
        let component = GetObjectProperty(this, propertyKey) as Component

        this.components.set(numId, component)
      }
    })
  }

  setParent(
    parentComponent: Component,
    parentNumId: number,
    parentPropertyKey?: string
  ) {
    this.parentComponent = parentComponent
    this.parentNumId = parentNumId

    if (parentPropertyKey) {
      this.parentPropertyKey = parentPropertyKey
    }
  }

  getParentComponent() {
    return this.parentComponent
  }

  getParentNumId() {
    return this.parentNumId
  }

  getParentPropertyKey() {
    return this.parentPropertyKey
  }

  loadCompleteStateBKP(state: any[]) {
    if (state[0] !== ComponentChangeType.complete) {
      return
    }

    const valuesSize = state[1]

    for (let i = 1; i < state.length; i += 3) {
      const key = state[i]
      const numId = state[i + 1]
      const value = state[i + 2]

      const isComponent = i >= 1 + valuesSize * 3

      if (isComponent) {
        //@ts-ignore
        const component = this[key]
        this.components.set(numId, component)
        component.setParent(this, key, numId)

        component.loadCompleteState(value)
      } else {
        this.properties.set(numId, key)
        //@ts-ignore
        this[key] = value

        this.emitOn(key, value)
      }
    }
  }

  loadCompleteState(state: any[]) {
    if (state[0] !== ComponentChangeType.complete) {
      return
    }

    for (let i = 1; i < state.length; i += 2) {
      const numId = state[i]
      const data = state[i + 1]

      const isComponent = this.components.has(numId)

      if (isComponent) {
        const component = this.components.get(numId)
        component.setParent(this, numId)

        component.loadCompleteState(data)
      } else {
        const key = this.properties.get(numId)
        //@ts-ignore
        this[key] = data

        this.emitOn(key, data)
      }
    }
  }

  on(key: string, callback: (value: any) => void) {
    if (!this.listeners.has(key)) {
      this.listeners.set(key, [])
    }

    this.listeners.get(key).push(callback)
  }

  private emitOn = (key: string, value: any) => {
    if (!this.listeners.has(key)) {
      return
    }

    const listeners = this.listeners.get(key)

    for (const listener of listeners) {
      listener(value)
    }
  }

  applyRecentChanges(changes: any[]) {
    if (changes[0] === ComponentChangeType.complete) {
      this.loadCompleteState(changes)
      return
    }

    if (changes[0] !== ComponentChangeType.partial) {
      return
    }

    for (let i = 1; i < changes.length; i += 2) {
      const numId = changes[i]
      const data = changes[i + 1]

      const isComponent = this.components.has(numId)

      if (isComponent) {
        const component = this.components.get(numId)
        component.applyRecentChanges(data)
      } else if (this.properties.has(numId)) {
        const key = this.properties.get(numId)
        //@ts-ignore
        this[key] = data

        this.emitOn(key, data)
      }
    }
  }

  applyRecentChangesBKP(data: any[]) {
    if (data[0] === ComponentChangeType.complete) {
      this.loadCompleteState(data)
      return
    }

    if (data[0] !== ComponentChangeType.partial) {
      return
    }

    for (let i = 1; i < data.length; i += 2) {
      const numId = data[i]
      const value = data[i + 1]

      if (this.components.has(numId)) {
        const component = this.components.get(numId)
        component.applyRecentChanges(value)
      } else if (this.properties.has(numId)) {
        const key = this.properties.get(numId)
        const value = data[i + 1]

        SetObjectProperty(this, key, value)
        this.emitOn(key, value)
      }
    }
  }
}
