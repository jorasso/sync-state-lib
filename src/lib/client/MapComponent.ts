import { ComponentChangeType } from "../server/Component"
import { Component } from "./Component"

type AllowedTypes = number | string | boolean

enum MapOperations {
  add = 0,
  remove = 1,
  update = 2,
}

export class MapComponent<T extends AllowedTypes> extends Component {
  map: Map<string, T> = new Map()

  onAddListeners: ((key: string, value: T) => void)[] = []
  onUpdateListeners: ((key: string, value: T) => void)[] = []
  onRemoveListeners: ((key: string) => void)[] = []

  get(key: string) {
    return this.map.get(key)
  }

  has(key: string) {
    return this.map.has(key)
  }

  onAdd(listener: (key: string, value: T) => void) {
    this.onAddListeners.push(listener)
  }

  emitOnAdd(key: string, value: T) {
    for (const listener of this.onAddListeners) {
      listener(key, value)
    }
  }

  onUpdate(listener: (key: string, value: T) => void) {
    this.onUpdateListeners.push(listener)
  }

  emitOnUpdate(key: string, value: T) {
    for (const listener of this.onUpdateListeners) {
      listener(key, value)
    }
  }

  onRemove(listener: (key: string) => void) {
    this.onRemoveListeners.push(listener)
  }

  emitOnRemove(key: string) {
    for (const listener of this.onRemoveListeners) {
      listener(key)
    }
  }

  size() {
    return this.map.size
  }

  loadCompleteState(state: any[]) {
    this.map.clear()
    this.properties.clear()

    for (let i = 2; i < state.length; i += 3) {
      const key = state[i]
      const numId = state[i + 1]
      const value = state[i + 2]

      this.properties.set(numId, key)

      this.map.set(key, value)

      this.emitOnAdd(key, value)
    }
  }

  applyRecentChanges(data: any[]): void {
    if (data[0] === ComponentChangeType.complete) {
      this.loadCompleteState(data)
      return
    }

    if (data[0] !== ComponentChangeType.partial) {
      return
    }

    for (let i = 1; i < data.length; i += 1) {
      const operation = data[i]

      if (operation === MapOperations.add) {
        const key = data[i + 1]
        const numId = data[i + 2]
        const value = data[i + 3]

        this.properties.set(numId, key)
        this.map.set(key, value)

        this.emitOnAdd(key, value)

        i += 3
      } else if (operation === MapOperations.remove) {
        const numId = data[i + 1]
        const key = this.properties.get(numId)

        this.map.delete(key)
        this.properties.delete(numId)

        this.emitOnRemove(key)

        i += 1
      } else if (operation === MapOperations.update) {
        const numId = data[i + 1]
        const value = data[i + 2]

        const key = this.properties.get(numId)

        this.map.set(key, value)

        this.emitOnUpdate(key, value)

        i += 1
      }
    }
  }
}
