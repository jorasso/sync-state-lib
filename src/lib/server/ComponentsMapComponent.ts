import { Component, ComponentChangeType } from "./Component"

enum MapOperations {
  add = 0,
  remove = 1,
  update = 2,
  replace = 3,
}

export class ComponentsMapComponent<T extends Component> extends Component {
  private map = new Map<string, T>()

  // Just to have the same signature as client's ComponentsMapComponent
  constructor(_: { new (): T }) {
    super()
  }

  size() {
    return this.map.size
  }

  set(key: string, component: T) {
    const currentComponent = this.map.get(key)

    if (currentComponent === component) {
      return
    }

    this.addToComponents(key, component, currentComponent?.getParentNumId())

    const numId = component.getParentNumId()

    this.addChildComponentPropertyChange(numId, currentComponent, component, {
      key,
    })

    currentComponent?.dispose()

    this.map.set(key, component)
  }

  get(key: string) {
    return this.map.get(key)
  }

  has(key: string) {
    return this.map.has(key)
  }

  delete(key: string) {
    const currentComponent = this.map.get(key)

    this.addChildComponentPropertyChange(
      currentComponent.getParentNumId(),
      currentComponent,
      undefined
    )

    currentComponent?.dispose()
    this.map.delete(key)
    this.removeFromComponents(currentComponent)
  }

  addToComponents = (
    key: string,
    component: Component,
    existingNumId: number
  ) => {
    const numId = existingNumId ?? this.getNumId()
    this.components.set(component, numId)
    component.setParent(this, key, numId)
  }

  removeFromComponents = (component: Component) => {
    this.components.delete(component)
  }

  getCompleteState() {
    const state: any[] = [ComponentChangeType.complete]

    state.push(this.map.size)

    this.map.forEach((component, key) => {
      state.push(key)
      state.push(this.components.get(component))

      state.push(component.getCompleteState())
    })

    return state
  }

  getRecentChanges() {
    const state: any[] = [ComponentChangeType.partial]

    this.childComponentsChanges.forEach((change, numId) => {
      if (change.to === undefined && change.from !== undefined) {
        state.push(MapOperations.remove)
        state.push(numId)

        return
      }

      if (change.from === undefined && change.to !== undefined) {
        state.push(MapOperations.add)
        state.push(change.meta.key)
        state.push(numId)
        state.push(change.to.getCompleteState())
        return
      }

      if (change.from !== change.to) {
        state.push(MapOperations.replace)
        state.push(numId)
        state.push(change.to.getCompleteState())
        return
      }

      const recentChanges = change.to.getRecentChanges()

      if (recentChanges.length === 0) {
        return
      }

      state.push(MapOperations.update)
      state.push(numId)
      state.push(recentChanges)
    })

    return state
  }
}
