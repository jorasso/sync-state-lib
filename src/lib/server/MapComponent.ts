import { Component, ComponentChangeType } from "./Component"

type AllowedTypes = number | string | boolean

enum MapOperations {
  add = 0,
  remove = 1,
  update = 2,
}

export class MapComponent<T extends AllowedTypes> extends Component {
  private map = new Map<string, T>()

  set(key: string, value: T) {
    this.addToProperties(key)

    const numId = this.properties.get(key)

    this.addPrimitivePropertyChange(numId, this.map.get(key), value, { key })

    this.map.set(key, value)
  }

  get(key: string) {
    return this.map.get(key)
  }

  has(key: string) {
    return this.map.has(key)
  }

  delete(key: string) {
    const numId = this.properties.get(key)
    this.addPrimitivePropertyChange(numId, this.map.get(key), undefined)
    this.map.delete(key)
    this.removeFromProperties(key)
  }

  addToProperties = (key: string) => {
    if (this.map.has(key)) {
      return
    }

    const numId = this.getNumId()

    this.properties.set(key, numId)
  }

  removeFromProperties = (key: string) => {
    this.properties.delete(key)
  }

  size() {
    return this.map.size
  }

  getCompleteState() {
    const state: any = [ComponentChangeType.complete]

    state.push(this.map.size)

    this.map.forEach((value, key) => {
      state.push(key)
      state.push(this.properties.get(key))

      state.push(value)
    })

    return state
  }

  getRecentChanges() {
    const state: any[] = [ComponentChangeType.partial]

    this.primitivePropertiesChanges.forEach((change, numId) => {
      if (change.from === change.to) {
        return
      }

      if (change.to === undefined) {
        state.push(MapOperations.remove)
        state.push(numId)

        return
      }

      if (change.from === undefined) {
        state.push(MapOperations.add)
        state.push(change.meta.key)
        state.push(numId)
        state.push(change.to)
        return
      }

      state.push(MapOperations.update)
      state.push(numId)
      state.push(change.to)
    })

    return state
  }
}
