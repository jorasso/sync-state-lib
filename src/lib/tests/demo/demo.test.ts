import { ComponentsMapComponent as ComponentsMapComponentServer } from "../../server/ComponentsMapComponent"
import { ComponentsMapComponent as ComponentsMapComponentClient } from "../../client/ComponentsMapComponent"

import {
  HeroServer,
  HeroClient,
  ActiveSlotServer,
  ActiveSlotClient,
} from "./components/DemoComponents"
import { expect, test, vi } from "vitest"

const GetData = () => {
  const server = new ComponentsMapComponentServer<HeroServer>(HeroServer)
  const client = new ComponentsMapComponentClient<HeroClient>(HeroClient)

  const heroServer = new HeroServer()
  heroServer.name = "Hero A"
  heroServer.x = 2000
  heroServer.y = 4000

  heroServer.inventory.items.set("A", 1)
  heroServer.inventory.items.set("B", 2)
  heroServer.inventory.items.set("C", 3)

  const activeSlotServer = new ActiveSlotServer()
  activeSlotServer.id = "AA"
  activeSlotServer.active = true

  heroServer.inventory.activeSlots.set("A", activeSlotServer)

  heroServer.aiming.angle = 90
  heroServer.aiming.active = true

  server.set("heroA", heroServer)

  const state = server.getCompleteState()

  client.loadCompleteState(state)

  server.resetRecentChanges()

  return { server, client, state }
}

test("complete state", () => {
  const { server, client, state } = GetData()

  // Check if server state is correct
  expect(server.has("heroA")).toBe(true)

  expect(server.get("heroA")).toBeDefined()

  expect(server.get("heroA").name).toBe("Hero A")
  expect(server.get("heroA").x).toBe(2000)
  expect(server.get("heroA").y).toBe(4000)

  expect(server.get("heroA").inventory.items.get("A")).toBe(1)
  expect(server.get("heroA").inventory.items.get("B")).toBe(2)
  expect(server.get("heroA").inventory.items.get("C")).toBe(3)

  expect(server.get("heroA").inventory.activeSlots.get("A")).toBeDefined()

  expect(server.get("heroA").inventory.activeSlots.get("A").id).toBe("AA")
  expect(server.get("heroA").inventory.activeSlots.get("A").active).toBe(true)

  expect(server.get("heroA").aiming.angle).toBe(90)
  expect(server.get("heroA").aiming.active).toBe(true)

  expect(server.size()).toBe(1)

  // Check if full state data is correct
  expect(state).toMatchSnapshot()

  // Check if client state is the same as the server state
  expect(client.has("heroA")).toBe(true)

  expect(client.get("heroA")).toBeDefined()

  expect(client.get("heroA").name).toBe("Hero A")
  expect(client.get("heroA").x).toBe(2000)
  expect(client.get("heroA").y).toBe(4000)

  expect(client.get("heroA").inventory.items.get("A")).toBe(1)
  expect(client.get("heroA").inventory.items.get("B")).toBe(2)
  expect(client.get("heroA").inventory.items.get("C")).toBe(3)

  expect(client.get("heroA").inventory.activeSlots.get("A")).toBeDefined()

  expect(client.get("heroA").inventory.activeSlots.get("A").id).toBe("AA")
  expect(client.get("heroA").inventory.activeSlots.get("A").active).toBe(true)

  expect(client.get("heroA").aiming.angle).toBe(90)
  expect(client.get("heroA").aiming.active).toBe(true)

  expect(client.size()).toBe(1)
})

test("callbacks for complete state", () => {
  const { client, state } = GetData()

  // Create callbacks for the client
  const addCallback = vi.fn()

  const nameUpdateCallback = vi.fn()
  const xUpdateCallback = vi.fn()
  const yUpdateCallback = vi.fn()
  const inventoryItemsAddCallback = vi.fn()
  const activeSlotsAddCallback = vi.fn()

  const idUpdateCallback = vi.fn()
  const activeUpdateCallback = vi.fn()

  const aimingAngleUpdateCallback = vi.fn()
  const aimingActiveUpdateCallback = vi.fn()

  client.onAdd((key: string, hero: HeroClient) => {
    addCallback(key, hero)

    // Create callbacks for the hero
    hero.on("name", nameUpdateCallback)

    hero.on("x", xUpdateCallback)

    hero.on("y", yUpdateCallback)

    hero.inventory.items.onAdd(inventoryItemsAddCallback)

    hero.inventory.activeSlots.onAdd(
      (key: string, activeSlot: ActiveSlotClient) => {
        activeSlotsAddCallback(key, activeSlot)

        activeSlot.on("id", idUpdateCallback)

        activeSlot.on("active", activeUpdateCallback)
      }
    )

    hero.aiming.on("angle", aimingAngleUpdateCallback)

    hero.aiming.on("active", aimingActiveUpdateCallback)
  })

  // It should reset client state and call all callbacks
  client.loadCompleteState(state)

  // Check if the callbacks are called with the correct values
  expect(addCallback).toHaveBeenCalledTimes(1)
  expect(addCallback).toHaveBeenCalledWith("heroA", client.get("heroA"))

  expect(nameUpdateCallback).toHaveBeenCalledTimes(1)
  expect(nameUpdateCallback).toHaveBeenCalledWith("Hero A")

  expect(xUpdateCallback).toHaveBeenCalledTimes(1)
  expect(xUpdateCallback).toHaveBeenCalledWith(2000)

  expect(yUpdateCallback).toHaveBeenCalledTimes(1)
  expect(yUpdateCallback).toHaveBeenCalledWith(4000)

  expect(inventoryItemsAddCallback).toHaveBeenCalledTimes(3)
  expect(inventoryItemsAddCallback).toHaveBeenCalledWith("A", 1)
  expect(inventoryItemsAddCallback).toHaveBeenCalledWith("B", 2)
  expect(inventoryItemsAddCallback).toHaveBeenCalledWith("C", 3)

  expect(activeSlotsAddCallback).toHaveBeenCalledTimes(1)
  expect(activeSlotsAddCallback).toHaveBeenCalledWith(
    "A",
    client.get("heroA").inventory.activeSlots.get("A")
  )

  expect(idUpdateCallback).toHaveBeenCalledTimes(1)
  expect(idUpdateCallback).toHaveBeenCalledWith("AA")

  expect(activeUpdateCallback).toHaveBeenCalledTimes(1)
  expect(activeUpdateCallback).toHaveBeenCalledWith(true)

  expect(aimingAngleUpdateCallback).toHaveBeenCalledTimes(1)
  expect(aimingAngleUpdateCallback).toHaveBeenCalledWith(90)

  expect(aimingActiveUpdateCallback).toHaveBeenCalledTimes(1)
  expect(aimingActiveUpdateCallback).toHaveBeenCalledWith(true)
})

test("run and aim", () => {
  const { server, client } = GetData()

  // Update the server state
  server.get("heroA").x = 3000
  server.get("heroA").y = 5000

  server.get("heroA").aiming.angle = 180

  // Check if the server state is correct
  expect(server.get("heroA").x).toBe(3000)
  expect(server.get("heroA").y).toBe(5000)

  expect(server.get("heroA").aiming.angle).toBe(180)

  // Send changes to the client
  const changes = server.getRecentChanges()
  server.resetRecentChanges()

  client.applyRecentChanges(changes)

  // Check if the client state is the same as the server state
  expect(client.get("heroA").x).toBe(server.get("heroA").x)
  expect(client.get("heroA").y).toBe(server.get("heroA").y)

  expect(client.get("heroA").aiming.angle).toBe(
    server.get("heroA").aiming.angle
  )
})

test("collect items", () => {
  const { server, client } = GetData()

  // Update the server state
  server.get("heroA").inventory.items.set("D", 4)

  // Check if the server state is correct
  expect(server.get("heroA").inventory.items.get("D")).toBe(4)
  expect(server.get("heroA").inventory.items.size()).toBe(4)

  // Send changes to the client
  const changes = server.getRecentChanges()
  server.resetRecentChanges()

  expect(changes).toMatchSnapshot()

  client.applyRecentChanges(changes)

  // Check if the client state is the same as the server state
  expect(client.get("heroA").inventory.items.get("D")).toBe(
    server.get("heroA").inventory.items.get("D")
  )
  expect(client.get("heroA").inventory.items.size()).toBe(
    server.get("heroA").inventory.items.size()
  )
})

test("drop items", () => {
  const { server, client } = GetData()

  // Update the server state
  server.get("heroA").inventory.items.delete("A")

  // Check if the server state is correct
  expect(server.get("heroA").inventory.items.get("A")).toBeUndefined()
  expect(server.get("heroA").inventory.items.size()).toBe(2)

  // Send changes to the client
  const changes = server.getRecentChanges()
  server.resetRecentChanges()

  client.applyRecentChanges(changes)

  // Check if the client state is the same as the server state
  expect(client.get("heroA").inventory.items.get("A")).toBeUndefined()
  expect(client.get("heroA").inventory.items.size()).toBe(2)
})

test("drop one item", () => {
  const { server, client } = GetData()

  // Update the server state
  server.get("heroA").inventory.items.set("C", 1)

  // Check if the server state is correct
  expect(server.get("heroA").inventory.items.get("C")).toBe(1)
  expect(server.get("heroA").inventory.items.size()).toBe(3)

  // Send changes to the client
  const changes = server.getRecentChanges()
  server.resetRecentChanges()

  client.applyRecentChanges(changes)

  // Check if the client state is the same as the server state
  expect(client.get("heroA").inventory.items.get("C")).toBe(1)
  expect(client.get("heroA").inventory.items.size()).toBe(3)
})

test("collect new active slot", () => {
  const { server, client } = GetData()

  // Update the server state
  const activeSlotServer = new ActiveSlotServer()
  activeSlotServer.id = "ZZZ"
  activeSlotServer.active = false

  server.get("heroA").inventory.activeSlots.set("ZZZ", activeSlotServer)

  // Check if the server state is correct
  expect(server.get("heroA").inventory.activeSlots.get("ZZZ")).toBeDefined()
  expect(server.get("heroA").inventory.activeSlots.size()).toBe(2)

  // Send changes to the client
  const changes = server.getRecentChanges()
  server.resetRecentChanges()

  client.applyRecentChanges(changes)

  // Check if the client state is the same as the server state
  expect(client.get("heroA").inventory.activeSlots.get("ZZZ")).toBeDefined()
  expect(client.get("heroA").inventory.activeSlots.size()).toBe(2)
})
