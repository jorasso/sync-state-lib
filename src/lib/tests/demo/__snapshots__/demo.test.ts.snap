// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`collect items 1`] = `
[
  2,
  2,
  0,
  [
    2,
    1,
    [
      2,
      1,
      [
        2,
        0,
        "D",
        3,
        4,
      ],
    ],
  ],
]
`;

exports[`complete state 1`] = `
[
  1,
  1,
  "heroA",
  0,
  [
    1,
    2,
    "Hero A",
    3,
    2000,
    4,
    4000,
    1,
    [
      1,
      1,
      [
        1,
        3,
        "A",
        0,
        1,
        "B",
        1,
        2,
        "C",
        2,
        3,
      ],
      0,
      [
        1,
        1,
        "A",
        0,
        [
          1,
          1,
          "AA",
          0,
          true,
        ],
      ],
    ],
    0,
    [
      1,
      1,
      90,
      0,
      true,
    ],
  ],
]
`;
